From 4c645e243cc01bfe67ba16b71f7a57a8bfa2b440 Mon Sep 17 00:00:00 2001
Message-ID: <4c645e243cc01bfe67ba16b71f7a57a8bfa2b440.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Thu, 16 Apr 2026 19:30:46 +0200
Subject: [RFC PATCH 18/28] virtio-msg: latch transport capabilities during
 pre-plug

Add a transport capability snapshot API and latch its values during
virtio-msg child pre-plug.

Fail pre-plug when the transport is not ready and apply the latched
extra host features to each child device.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg-bus.c         | 22 ++++++++++++++++
 hw/virtio/virtio-msg.c             | 40 +++++++++++++++++++++++++++++-
 include/hw/virtio/virtio-msg-bus.h | 20 ++++++++++++---
 include/hw/virtio/virtio-msg.h     |  4 +++
 4 files changed, 81 insertions(+), 5 deletions(-)

diff --git a/hw/virtio/virtio-msg-bus.c b/hw/virtio/virtio-msg-bus.c
index 4e968eb0323d..46cdca1cad53 100644
--- a/hw/virtio/virtio-msg-bus.c
+++ b/hw/virtio/virtio-msg-bus.c
@@ -71,6 +71,28 @@ int virtio_msg_bus_send_must_deliver(BusState *bus, VirtIOMSG *msg_req)
     return virtio_msg_bus_send(bus, msg_req);
 }
 
+void virtio_msg_bus_get_transport_caps(BusState *bus,
+                                       VirtIOMSGTransportCaps *caps)
+{
+    VirtIOMSGBusDevice *bd = virtio_msg_bus_get_device(bus);
+    VirtIOMSGBusDeviceClass *bdc;
+
+    caps->dma_as = &address_space_memory;
+    caps->extra_features = 0;
+    caps->ready = false;
+
+    if (!bd) {
+        return;
+    }
+
+    bdc = VIRTIO_MSG_BUS_DEVICE_CLASS(object_get_class(OBJECT(bd)));
+    if (bdc->get_transport_caps) {
+        bdc->get_transport_caps(bd, caps);
+        return;
+    }
+    caps->ready = true;
+}
+
 static void virtio_msg_bus_class_init(ObjectClass *klass, const void *data)
 {
     BusClass *bc = BUS_CLASS(klass);
diff --git a/hw/virtio/virtio-msg.c b/hw/virtio/virtio-msg.c
index f67a245289d8..6a6b6d59525c 100644
--- a/hw/virtio/virtio-msg.c
+++ b/hw/virtio/virtio-msg.c
@@ -847,17 +847,52 @@ assign_error:
     return r;
 }
 
+static void virtio_msg_reset_latched_caps(VirtIOMSGProxy *s)
+{
+    s->latched_caps.dma_as = &address_space_memory;
+    s->latched_caps.extra_features = 0;
+    s->latched_caps.ready = false;
+    s->latched_caps_valid = false;
+}
+
+static bool virtio_msg_latch_transport_caps(VirtIOMSGProxy *s, Error **errp)
+{
+    VirtIOMSGTransportCaps caps;
+
+    if (s->latched_caps_valid) {
+        return true;
+    }
+
+    virtio_msg_bus_get_transport_caps(&s->msg_bus, &caps);
+    if (!caps.ready) {
+        error_setg(errp, "virtio-msg transport is not ready during pre-plug");
+        return false;
+    }
+
+    s->latched_caps = caps;
+    s->latched_caps_valid = true;
+    return true;
+}
+
 static void virtio_msg_pre_plugged(DeviceState *d, Error **errp)
 {
     VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(d);
     VirtIOMSGProxy *s = VIRTIO_MSG(mdev->proxy);
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, mdev->dev_num, __func__);
+    VirtIODevice *vdev = virtio_msg_vdev(s, mdev->dev_num);
+    uint64_t extra_features;
 
     if (!vdev) {
         return;
     }
 
+    if (!virtio_msg_latch_transport_caps(s, errp)) {
+        return;
+    }
+
+    extra_features = s->latched_caps.extra_features;
+
     virtio_add_feature(&vdev->host_features, VIRTIO_F_VERSION_1);
+    vdev->host_features |= extra_features;
 }
 
 static VirtIOMSGProxy *virtio_msg_resolve_proxy(DeviceState *d)
@@ -895,6 +930,9 @@ static void virtio_msg_realize(DeviceState *d, Error **errp)
     VirtIOMSGProxy *s = VIRTIO_MSG(d);
     Object *o = OBJECT(d);
     int i;
+    (void)errp;
+
+    virtio_msg_reset_latched_caps(s);
 
     for (i = 0; i < ARRAY_SIZE(s->devs); i++) {
         g_autofree char *outer_bus_name = g_strdup_printf("bus%d", i);
diff --git a/include/hw/virtio/virtio-msg-bus.h b/include/hw/virtio/virtio-msg-bus.h
index d85062d36c9f..cf401b931cfa 100644
--- a/include/hw/virtio/virtio-msg-bus.h
+++ b/include/hw/virtio/virtio-msg-bus.h
@@ -29,6 +29,12 @@ typedef struct VirtIOMSGBusPort {
     bool is_driver;
 } VirtIOMSGBusPort;
 
+typedef struct VirtIOMSGTransportCaps {
+    AddressSpace *dma_as;
+    uint64_t extra_features;
+    bool ready;
+} VirtIOMSGTransportCaps;
+
 struct VirtIOMSGBusDeviceClass {
     /*< private >*/
     DeviceClass parent_class;
@@ -55,6 +61,14 @@ struct VirtIOMSGBusDeviceClass {
      * delivered eventually.
      */
     int (*send_must_deliver)(VirtIOMSGBusDevice *bd, VirtIOMSG *msg_req);
+
+    /*
+     * Return a coherent transport capability snapshot for child pre-plug.
+     * If unset, dma_as remains address_space_memory.
+     */
+    void (*get_transport_caps)(VirtIOMSGBusDevice *bd,
+                               VirtIOMSGTransportCaps *caps);
+
 };
 
 typedef struct VirtIOMSGBusDevice {
@@ -104,8 +118,6 @@ virtio_msg_bus_receive(VirtIOMSGBusDevice *bd, VirtIOMSG *msg)
 int virtio_msg_bus_send(BusState *bus, VirtIOMSG *msg_req);
 int virtio_msg_bus_send_must_deliver(BusState *bus, VirtIOMSG *msg_req);
 
-static inline AddressSpace *virtio_msg_bus_get_remote_as(BusState *bus)
-{
-    return &address_space_memory;
-}
+void virtio_msg_bus_get_transport_caps(BusState *bus,
+                                       VirtIOMSGTransportCaps *caps);
 #endif
diff --git a/include/hw/virtio/virtio-msg.h b/include/hw/virtio/virtio-msg.h
index c25195b1bc55..9c855dd5690f 100644
--- a/include/hw/virtio/virtio-msg.h
+++ b/include/hw/virtio/virtio-msg.h
@@ -12,6 +12,7 @@
 
 #include "hw/core/sysbus.h"
 #include "hw/virtio/virtio-bus.h"
+#include "hw/virtio/virtio-msg-bus.h"
 
 #define TYPE_VIRTIO_MSG_PROXY_BUS "virtio-msg-proxy-bus"
 /* This is reusing the VirtioBusState typedef from TYPE_VIRTIO_BUS */
@@ -50,6 +51,9 @@ struct VirtIOMSGProxy {
     BusState devs_bus[VIRTIO_MSG_MAX_DEVS];
     VirtIOMSGDev devs[VIRTIO_MSG_MAX_DEVS];
 
+    VirtIOMSGTransportCaps latched_caps;
+    bool latched_caps_valid;
+
     /* virtio-msg-bus.  */
     BusState msg_bus;
 };
-- 
2.53.0

